viewer-prototype/src/browser/trace-server-status.ts

```diff
@@ -0,0 +1,101 @@
+/********************************************************************************
+ * Copyright (C) 2020 École Polytechnique de Montréal
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { ILogger } from '@theia/core/lib/common/logger';
+import { DefaultFrontendApplicationContribution } from '@theia/core/lib/browser/frontend-application';
+import { StatusBar, StatusBarAlignment } from '@theia/core/lib/browser/status-bar/status-bar';
+import { Disposable, DisposableCollection } from '@theia/core/lib//common';
+import { ConnectionStatusService, ConnectionStatus, AbstractConnectionStatusService } from '@theia/core/lib/browser/connection-status-service';
+import { TspClient } from 'tsp-typescript-client/lib/protocol/tsp-client';
+
+@injectable()
+export class TraceServerConnectionStatusService extends AbstractConnectionStatusService {
+
+    private scheduledPing: number | undefined;
+
+    private constructor(
+        @inject(TspClient) private tspClient: TspClient
+    ) {
+        super();
+    }
+
+    @postConstruct()
+    protected init(): void {
+        this.schedulePing();
+    }
+
+    protected schedulePing(): void {
+        if (this.scheduledPing) {
+            this.clearTimeout(this.scheduledPing);
+        }
+        this.scheduledPing = this.setTimeout(async () => {
+            try {
+                await this.tspClient.fetchExperiments();
+                this.updateStatus(true);
+            } catch (e) {
+                this.logger.trace(e);
+            }
+            this.schedulePing();
+        }, this.options.offlineTimeout * 0.8);
```

Comment:
This means a call every like 4.5 seconds... Maybe too much? this.options.offlineTimeout *2.0 maybe?

Reply:
*2.0 makes the trace server unavailable one in 2 requests, I guess it compares with the timeout. Having a ping on the server will make make it more efficient